src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#include "../LifetimeExtender.h"
+#include <NovelRT.Interop/NrtErrorHandling.h>
+#include <NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.h>
+#include <NovelRT/PluginManagement/PluginManagement.h>
+
+using namespace NovelRT::PluginManagement;
+using namespace NovelRT::Interop;
+
+NrtDefaultPluginSelectorHandle Nrt_DefaultPluginSelector_Create()
+{
+    auto pluginSelector = new DefaultPluginSelector();
+    return reinterpret_cast<NrtDefaultPluginSelectorHandle>(pluginSelector);
+}
+
+NrtResult Nrt_DefaultPluginSelector_GetDefaultResourceManagementPluginForCurrentPlatform(
+    NrtDefaultPluginSelectorHandle pluginSelector,
+    NrtIResourceManagementPluginProviderHandle* outputResourceManagementPlugin)
+{
+    if (pluginSelector == nullptr)
+    {
+        Nrt_setErrIsNullInstanceProvidedInternal();
+        return NRT_FAILURE_NULL_INSTANCE_PROVIDED;
+    }
+
+    if (outputResourceManagementPlugin == nullptr)
+    {
+        Nrt_setErrIsNullArgProvidedInternal();
+        return NRT_FAILURE_NULL_ARGUMENT_PROVIDED;
+    }
+
+    auto&& resourcePlugin = reinterpret_cast<DefaultPluginSelector*>(pluginSelector)
+                                ->GetDefaultPluginTypeOnCurrentPlatformFor<IResourceManagementPluginProvider>();
+    *outputResourceManagementPlugin =
+        reinterpret_cast<NrtIResourceManagementPluginProviderHandle>(resourcePlugin.get());
+
+    Lifetime::KeepAlive(resourcePlugin);
+
+    return NRT_SUCCESS;
+}
+NrtResult Nrt_DefaultPluginSelector_GetDefaultGraphicsPluginForCurrentPlatform(
+    NrtDefaultPluginSelectorHandle pluginSelector,
+    NrtIGraphicsPluginProviderHandle* outputGraphicsPlugin)
+{
+    if (pluginSelector == nullptr)
+    {
+        Nrt_setErrIsNullInstanceProvidedInternal();
+        return NRT_FAILURE_NULL_INSTANCE_PROVIDED;
+    }
+
+    if (outputGraphicsPlugin == nullptr)
+    {
+        Nrt_setErrIsNullArgProvidedInternal();
+        return NRT_FAILURE_NULL_ARGUMENT_PROVIDED;
+    }
+
+    auto&& graphicsPlugin = reinterpret_cast<DefaultPluginSelector*>(pluginSelector)
```

Comment:
Out of curiosity, what does the `auto&&` achieve here as opposed to just `auto`?

Reply:
Nothing much, using `auto` would lead to the same behavior. I liked `auto&&` because it sort of says that "this value is temporary and will be moved".